server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesUpdateProcessor.java

```diff
@@ -157,26 +166,40 @@ private void handleUpdate(MailboxId mailboxId, MailboxUpdateRequest updateReques
         }
    }
 
+    private void assertNotSharedOutboxOrDraftMailbox(Mailbox mailbox, MailboxUpdateRequest updateRequest) {
+        mailbox.getRole().ifPresent(role ->
+            Preconditions.checkArgument(!isShardOutboxOrDraft(role, updateRequest.getSharedWith()), "Cannot share 'Outbox' or 'Draft'"));
```

Comment:
I find it weird to get a Precondition inside a lambda. Moreover I think we can get two preconditions: one for Outbox, one for draft : the error messages would be more specific.

```
Preconditions.checkArguments(isSetSharedWithOn(updateRequest, Role.DRAFTS), "Can not share 'Drafts'");
Preconditions.checkArguments(isSetSharedWithOn(updateRequest, Role.OUTBOX), "Can not share 'Drafts'");
```

And : 

```
public boolean isSetSharedWithOn(UpdateRequest updateRequest, Role role) {
    return updateRequest.getSharedWhith.isPresent() && mailbox.hasRole(role);
}
```

We also need to add a `hasRole` method on Mailbox (with @JsonIgnore )